apm-agent-core/src/test/java/co/elastic/apm/agent/bci/modules/ModuleOpenerTest.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.bci.modules;
+
+import net.bytebuddy.agent.ByteBuddyAgent;
+import org.apache.commons.compress.utils.IOUtils;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.EnabledForJreRange;
+import org.junit.jupiter.api.condition.JRE;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.instrument.Instrumentation;
+import java.lang.reflect.Method;
+import java.util.Collections;
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class ModuleOpenerTest {
+
+    @Test
+    @EnabledForJreRange(min = JRE.JAVA_17)
+    public void verifyModuleOpening() throws Exception {
+        String illegalAccessorName = IllegalAccessor.class.getName();
+
+        ClassLoader cl1 = new ChildFirstCopyClassloader(getClass().getClassLoader(), illegalAccessorName);
+        Method illegal1 = cl1.loadClass(illegalAccessorName).getMethod("doAccess");
+        ClassLoader cl2 = new ChildFirstCopyClassloader(getClass().getClassLoader(), illegalAccessorName);
+        Method illegal2 = cl2.loadClass(illegalAccessorName).getMethod("doAccess");
+
+        assertThatThrownBy(() -> illegal1.invoke(null))
+            .hasCauseInstanceOf(IllegalAccessException.class);
+
+        Instrumentation instr = ByteBuddyAgent.install();
+        Class<?> classFromModuleToOpen = Class.forName("com.sun.jndi.ldap.LdapResult");
+        ModuleOpener.getInstance().openModuleTo(instr, classFromModuleToOpen, cl2, Collections.singleton("com.sun.jndi.ldap"));
+
+        //Access should be legal now and not throw any exception
+        illegal2.invoke(null);
+    }
+
+    public static class IllegalAccessor {
+
+        public static void doAccess() throws Exception {
+            //com.sun.jndi.ldap.LdapResult is not accessible according to module rules
+            Class.forName("com.sun.jndi.ldap.LdapResult").getConstructor().newInstance();
+        }
+
+    }
+
+    private static class ChildFirstCopyClassloader extends ClassLoader {
```

Comment:
Would [ChildFirstURLClassLoader](https://github.com/elastic/apm-agent-java/blob/9575d839280b71e6ef771be45e5436d241b68c04/apm-agent-core/src/test/java/co/elastic/test/ChildFirstURLClassLoader.java#L39) be useful here?

Reply:
I actually looked at it and the `TestClassWithDependenciesRunner`: I decided against it because it requires extracting the bytecode of the class into a temporary jar.
Instead I used this implementation because:
 * It does not involve disk access coding
 * The code is not longer and complexer than reusing the `ChildFirstURLClassLoader` because it required packaging the target class in a temp jar first